### PR TITLE
adds model_name option in ConfigDict

### LINKED ISSUE #6304: Allowing overriding of model names in `get_model_name_map`
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

I'd like to be able override the model names generated by `get_model_name_map()`.

In the case of conflicts, often the verbose version ([ref](https://github.com/pydantic/pydantic/blob/main/pydantic/v1/schema.py#L430)) of the model name is excessively verbose. 

I'm using FastAPI to generate an OpenAPI spec of the app I'm building, which ultimately is being used to generate a client library for my service. It uses pydantic's `get_model_name_map` to generate model names to create the json spec. Because of a number of models have conflicting names, I end up with generated model names that are frustratingly verbose. 

`service/clients/clientA.py`
```python
clientA.py

class User(BaseModel):
  ...
```

`service/clients/clientB.py`
```python
class User(BaseModel):
  ...
```

`service/main.py`
```python
from service.clients import clientA, clientB

clientA.User()
clientB.User()
```

Because of the collision on the `User` model, when `get_model_name_map()` is run, I end up with model names like `service__clients__clientA__User` and `service__clients__clientB__User`. What I'd like to have is `ClientAUser` and `ClientBUser`.

I think it would be useful to allow overrides of this long model name to allow user-defined model titles when there are collisions.

Imagine perhaps a configuration like this, where the generated names are taken directly from the model config:

`service/clients/clientA.py`
```python
clientA.py

class User(BaseModel):
  class Config:
    title = "ClientAUser"
  ...
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [X] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

Selected Assignee: @lig

### PR DESCRIPTION
**Pull request by romulocollopy**
## Change Summary
Adds `model_name` attribute to `ConfigDict`. This will set the `cls.__name__`

## Related issue number

Fix #6304 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by sydney-runkle** [comment]
@romulocollopy,

This generally looks good - would you want to rebase + open this PR for review?

**Comment by sydney-runkle** [comment]
@romulocollopy, is this ready for review now? If so, could you please move it out of draft state?

**Comment by romulocollopy** [comment]
> @romulocollopy, is this ready for review now? If so, could you please move it out of draft state?

@sydney-runkle  I changed some formatting accidentally, I will check the linters configuration and push again to open the PR

**Comment by sydney-runkle** [comment]
@romulocollopy, I think our contribution guidelines might be able to help with linter config: https://docs.pydantic.dev/latest/contributing/

File: pydantic/config.py [hunk 1]
```diff
@@ -33,6 +33,9 @@ class ConfigDict(TypedDict, total=False):
     title: str | None
     """The title for the generated JSON schema, defaults to the model's name"""
 
+    model_name: str | None
+    """If not `None` overrides the model names generated by `pydantic.v1.schema.get_model_name_map()`."""
```
Comment (sydney-runkle) [inline pydantic/config.py:37]:
```suggestion
    """Name assigned to `__name__` and used in the model's JSON schema"""
```

Comment (sydney-runkle) [reply pydantic/config.py:37]: We can omit the mentions of inner-workings of the code and instead give a more broad description here that'll help users decide if they want to set this field.

File: pydantic/v1/schema.py [hunk 2]
```diff
@@ -89,7 +89,8 @@
 default_prefix = '#/definitions/'
 default_ref_template = '#/definitions/{model}'
 
-TypeModelOrEnum = Union[Type['BaseModel'], Type[Enum]]
+T = TypeVar('T', bound='BaseModel')
+TypeModelOrEnum = Union[Type[T], Type[Enum]]
```
Comment (sydney-runkle) [inline pydantic/v1/schema.py:93]: Is this change necessary?

File: tests/test_config.py [hunk 3]
```diff
@@ -492,6 +493,66 @@ def my_function():
         pass
 
 
+def test_config_model_name() -> None:
```
Comment (sydney-runkle) [inline tests/test_config.py:496]: Perhaps I'm missing something here, but isn't the only thing that we really need to focus on testing just that the name of the model when assigned via `model_name` is correct? Why even bring in `get_model_name_map`?

You'll definitely want to run a test or two with model names not equivalent to what they would by default be.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Just a few comments / requests, looking good thus far!

File: pydantic/_internal/_model_construction.py [hunk 4]
```diff
@@ -88,6 +88,7 @@ def __new__(
             base_field_names, class_vars, base_private_attributes = mcs._collect_bases_data(bases)
 
             config_wrapper = ConfigWrapper.for_model(bases, namespace, kwargs)
+            cls_name = config_wrapper.config_dict.get('model_name') or cls_name
```
Comment (Viicos) [inline pydantic/_internal/_model_construction.py:91]: Would it be possible to have some kind of "backup" of the original class name if overridden by `config.model_name`? I'm making use of `cls.__name__` [here](https://github.com/pydantic/pydantic/pull/6563/files#diff-515831ae80b48a711193d4896331a746d0e6e2ff0116621b58a068e4d242acc2R70), and comparing it to AST-parsed file.

In fact it would be great to have this PR merged before https://github.com/pydantic/pydantic/pull/6563/, I will then be able to rebase on top of it and add this "backup" field

**Comment by sydney-runkle** [comment]
@romulocollopy,

Feel free to move this out of draft when you have a chance 👍

File: pydantic/v1/schema.py (hunk 2 above)
Comment (romulocollopy) [reply pydantic/v1/schema.py:93]: Hum... I think no, will remove

File: pydantic/_internal/_model_construction.py (hunk 4 above)
Comment (romulocollopy) [reply pydantic/_internal/_model_construction.py:91]: I think that if `cls.__name__` is used somewhere else, maybe it would be better to keep its original behaviour and create an extra attribute for the model_name that will explicitly interact with the `v1.schema` functions. What do you think?

File: tests/test_config.py (hunk 3 above)
Comment (romulocollopy) [reply tests/test_config.py:496]: The goal of this PR was to address https://github.com/pydantic/pydantic/issues/6304, but considering [this](https://github.com/pydantic/pydantic/pull/6814/files/090985bddef2c2d9462ca177ec768f3c77d08cc0#r1451543032) comment, maybe it is better to not touch the `cls.__name__` and change the behaviour of `get_model_name_map` by making it check a new attribute with the `model_name`.

I would rework a bit the implementation in this direction and keep this test. What do you think?

File: pydantic/_internal/_model_construction.py (hunk 4 above)
Comment (Viicos) [reply pydantic/_internal/_model_construction.py:91]: Yes this seems better: `__name__` is a Python implementation detail and people should expect it to be the same as defined in the source code (it might also be used by documentation tools). Perhaps you can define a specific attribute that will be used for schema generation

File: pydantic/v1/schema.py [hunk 5]
```diff
@@ -333,7 +333,10 @@ def get_model_name_map(unique_models: TypeModelSet) -> Dict[TypeModelOrEnum, str
     name_model_map = {}
     conflicting_names: Set[str] = set()
     for model in unique_models:
-        model_name = normalize_name(model.__name__)
+        try:
+            model_name = model.model_config["model_name"]
+        except AttributeError:
+            model_name = normalize_name(model.__name__)
```
Comment (romulocollopy) [inline pydantic/v1/schema.py:339]: I could try something like this if it is preferable:
```suggestion  
        model_name = (
            model.model_config["model_name"] if not isinstance(model, Enum)
            else normalize_name(model.__name__)
        )
```

File: pydantic/config.py [hunk 6]
```diff
@@ -177,22 +180,22 @@ class User(BaseModel):
...
         some_enum: SomeEnum
-        another_enum: Optional[SomeEnum] = Field(default=SomeEnum.FOO, validate_default=True)
-
+        another_enum: Optional[SomeEnum] = Field(
+            default=SomeEnum.FOO, validate_default=True
+        )
 
     model1 = SomeModel(some_enum=SomeEnum.BAR)
     print(model1.model_dump())
+    #> {'some_enum': 'bar', 'another_enum': 'foo'}
```
Comment (romulocollopy) [inline pydantic/config.py:198]: I'll revert the formating. I have to turn off my black when saving.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
@romulocollopy,

Thanks for your work on this. I know that the initial issue was filed with V1 in mind, but I think we primarily need to focus on V2 here. We're no longer making changes to V1 unless they're security oriented - we're not adding new features.

I think the changes you had to the `_model_construction.py` file with this addition were on the right track:
```py
cls_name = config_wrapper.config_dict.get('model_name') or cls_name
```

@Viicos, in order for this to be helpful for your #6563 PR, do you need any changes to V1 made?

Let me know if you have any follow up questions.

**Comment by Viicos** [comment]
> I think the changes you had to the `_model_construction.py` file with this addition were on the right track:
> 
> ```python
> cls_name = config_wrapper.config_dict.get('model_name') or cls_name
> ```
> 
> @Viicos, in order for this to be helpful for your #6563 PR, do you need any changes to V1 made?
> 
> Let me know if you have any follow up questions.

If the class name gets changed, I'll have to update my PR once this is merged. But as said [here](https://github.com/pydantic/pydantic/pull/6814#discussion_r1452700846), it might be better to not set the actual class name to something else, and instead define a custom attribute that will be used when generating the JSON schema. If we go with the later, then no changes will be required in my PR

**Comment by sydney-runkle** [comment]
> If the class name gets changed, I'll have to update my PR once this is merged. But as said https://github.com/pydantic/pydantic/pull/6814#discussion_r1452700846, it might be better to not set the actual class name to something else, and instead define a custom attribute that will be used when generating the JSON schema. If we go with the later, then no changes will be required in my PR

Ah yes, this sounds good.

**Comment by Viicos** [comment]
Aaah, now I'm a bit confused, what use case having `model_name` is trying to solve where `title` is not enough (in v2)?

Looking at the code base, I see it is being used in `GenerateSchema._model_schema`.

I'm also a bit concerned about having the `__name__` attribute changed, e.g. what happens in this function?

https://github.com/pydantic/pydantic/blob/5de67f7c4f775b5319af9be157f9c7b68995eb1e/pydantic/_internal/_typing_extra.py#L195-L198

**Comment by sydney-runkle** [comment]
@Viicos,

I agree, we shouldn't change the `__name__` attribute (as I had originally suggested).

I think a better solution here will be supporting a hook for changing the defs ref for a class. @alexmojaki seemed to be on the right track here: https://github.com/pydantic/pydantic/issues/6304#issuecomment-1763458278.

More specifically, having some sort of dunder value that specifies what to use in the `get_defs_ref` function of the `GenerateJsonSchema` class will probably be the best approach.

**Comment by sydney-runkle** [comment]
Tidying up PRs -- closing this as stale for now, but we can certainly reopen if it makes sense down the line!
